storage/rocksdb/ha_rocksdb.cc

```diff
@@ -344,6 +345,89 @@ static std::string rdb_normalize_dir(std::string dir) {
+    THD *const thd MY_ATTRIBUTE((__unused__)),
+    struct st_mysql_sys_var *const var MY_ATTRIBUTE((__unused__)),
+    void *const save MY_ATTRIBUTE((__unused__)),
+    struct st_mysql_value *const value) {
+  char buf[FN_REFLEN];
+  int len = sizeof(buf);
+  const char *const trace_opt_str_raw = value->val_str(value, buf, &len);
+  rocksdb::Status s;
+  if (trace_opt_str_raw && rdb != nullptr) {
+    int rc __attribute__((__unused__));
+    std::string trace_opt_str(trace_opt_str_raw);
+    if (trace_opt_str.empty()) {
+      // End tracing block cache accesses.
+      sql_print_information("RocksDB: Stop tracing block cache accesses.\n");
+      s = rdb->EndBlockCacheTrace();
+      if (!s.ok()) {
+        rc = ha_rocksdb::rdb_error_to_mysql(s);
+        return HA_EXIT_FAILURE;
+      }
+      return HA_EXIT_SUCCESS;
+    }
+
+    // Start tracing block cache accesses.
+    std::stringstream ss(trace_opt_str);
+    std::vector<std::string> trace_opts_strs;
+    while (ss.good()) {
+      std::string substr;
+      getline(ss, substr, ':');
+      trace_opts_strs.push_back(substr);
+    }
+    if (trace_opts_strs.size() != 3) {
+      sql_print_information(
+          "RocksDB: Failed to parse trace option string: %s. The correct "
+          "format is sampling_frequency:max_trace_file_size:trace_file_path. "
+          "sampling_frequency and max_trace_file_size are positive integers. "
+          "The block accesses are saved to the trace_file_path.\n",
+          trace_opt_str.c_str());
+      return HA_EXIT_FAILURE;
+    }
+    rocksdb::TraceOptions trace_opt;
+    try {
+      trace_opt.sampling_frequency = std::stoull(trace_opts_strs[0]);
+      trace_opt.max_trace_file_size = std::stoull(trace_opts_strs[1]);
+    } catch (const std::exception& e) {
+      sql_print_information(
+          "RocksDB: Failed to parse trace option string: %s. The correct "
+          "format is sampling_frequency:max_trace_file_size:trace_file_path. "
+          "sampling_frequency and max_trace_file_size are positive integers. "
+          "The block accesses are saved to the trace_file_path.\n",
+          trace_opt_str.c_str());
+      return HA_EXIT_FAILURE;
+    }
+    const std::string trace_file_path = trace_opts_strs[2];
+    sql_print_information(
+        "RocksDB: Start tracing block cache accesses. Sampling frequency: %lu, "
+        "Maximum trace file size: %lu, Trace file path %s.\n",
+        trace_opt.sampling_frequency, trace_opt.max_trace_file_size,
+        trace_file_path.c_str());
+    std::unique_ptr<rocksdb::TraceWriter> trace_writer;
+    const rocksdb::EnvOptions env_option(rdb->GetDBOptions());
+    s = rocksdb::NewFileTraceWriter(rdb->GetEnv(), env_option, trace_file_path,
+                                    &trace_writer);
+    if (!s.ok()) {
+      rc = ha_rocksdb::rdb_error_to_mysql(s);
+      return HA_EXIT_FAILURE;
+    }
+    s = rdb->StartBlockCacheTrace(trace_opt, std::move(trace_writer));
```

Comment:
Is it valid to call StartBlockCacheTrace() twice in a row with different settings? Would it make more sense to force the state transition from on -> off and off -> on?